### PR TITLE
Warn for remotables not declared Remotable or Far

### LINKED ISSUE #2018: Empty objects should be data, Far should be marked as such
**Describe the bug**
The current specified behaviour of `@agoric/marshal` is to treat empty objects as `Remotable`.  This is handy for creating object handles which are `===` to themselves when returned to the vat that created them.    However, this convenience comes at a cost of inadvertently creating Presences when only an empty data object is intended.

**To Reproduce**
1. Send an empty object (`{}`) as an argument to a method defined in another vat (via `@agoric/marshal`)
2. `console.log` out the received argument.
3. Note that it displays as a Presence, not just an empty object.

**Expected behavior**
It would be clearer to have empty data objects treated as pass-by-copy.

**Platform Environment**
 - what OS are you using? what version of Node.js? n/a
 - is there anything special/unusual about your platform? n/a
 - what version of the Agoric-SDK are you using? latest master as of 2020-11-13

**Additional context**
See #804 for some discussion

As a migration step, requiring people to explicitly mark empty objects as either `Remotable` or `Data` would be useful to detect up-front when they haven't considered what the object is intended to be.  During this intermediate step, the `Data` marking could be applied automatically to empty objects that are received as pass-by-copy from another vat to prevent false positives (only empty objects constructed from within the vat would need explicit marking).

`Data` can be made optional in a later step when the default for unmarked empty objects is pass-by-copy.

### PR DESCRIPTION
**Pull request by erights**
A step towards #3525 

Deprecates implicit remotables. Makes more implicit remotables explicit.

### CODE REVIEW

**Comment by erights** [comment]
@michaelfig @warner I have two remaining failures:
https://github.com/Agoric/agoric-sdk/pull/3504/checks?check_run_id=3127742280
and
https://github.com/Agoric/agoric-sdk/pull/3504/checks?check_run_id=3127767756

which both seem to be about timeouts. But I suspect it is something else. Could you take a quick look and let me know what else comes to mind? Thanks.

**Comment by warner** [comment]
I don't know what's happening with those failures. I'm tempted to just push the "re-run all jobs" button and see if they go away. And/or if you rebase to current trunk and force-push, or add a commit, that will also trigger a rebuild.

**Comment by michaelfig** [comment]
> @michaelfig @warner I have two remaining failures:
> https://github.com/Agoric/agoric-sdk/pull/3504/checks?check_run_id=3127742280

https://github.com/Agoric/agoric-sdk/pull/3504/checks?check_run_id=3127742280#step:8:280 shows that the ag-solo is waiting on a notifier that never fires.

> https://github.com/Agoric/agoric-sdk/pull/3504/checks?check_run_id=3127767756

https://github.com/Agoric/agoric-sdk/pull/3504/checks?check_run_id=3127767756#step:7:1 could be the same thing.  Please run that test on your computer with `cd packages/solo && NOISY=true yarn test -v`

> which both seem to be about timeouts. But I suspect it is something else. Could you take a quick look and let me know what else comes to mind? Thanks.

**Comment by erights** [comment]
> Please run that test on your computer with `cd packages/solo && NOISY=true yarn test -v`

```
(1):solo(2018-markm-far-must-be-declared)$ NOISY=true yarn test -v
yarn run v1.22.5
$ ava -v

# connecting
.
'@agoric/install-ses' sniffed and found a 'LOCKDOWN_OPTIONS' environment variable

.
.
2021-07-22T22:17:00.695Z ag-solo: init: ag-solo initialized in /var/folders/g1/crzbwthn3n72s_8smhth9g_c0000gn/T//startsolo.83908
2021-07-22T22:17:00.695Z ag-solo: init: HTTP/WebSocket will listen on 127.0.0.1:7999
.
'@agoric/install-ses' sniffed and found a 'LOCKDOWN_OPTIONS' environment variable

.
.
'@agoric/install-ses' sniffed and found a 'LOCKDOWN_OPTIONS' environment variable

.
.
.
.
.
.
.
.
.
.
2021-07-22T22:17:13.682Z SwingSet: kernel: v5 online: xs-worker static name=bootstrap transcript entries replayed: 0
2021-07-22T22:17:13.695Z SwingSet: kernel: v7 online: local static name=comms transcript entries replayed: null
2021-07-22T22:17:13.851Z SwingSet: kernel: v2 online: xs-worker static name=http transcript entries replayed: 0
2021-07-22T22:17:13.998Z SwingSet: kernel: v3 online: xs-worker static name=network transcript entries replayed: 0
2021-07-22T22:17:14.155Z SwingSet: kernel: v1 online: xs-worker static name=spawner transcript entries replayed: 0
.
2021-07-22T22:17:14.290Z SwingSet: kernel: v9 online: xs-worker static name=timer transcript entries replayed: 0
2021-07-22T22:17:14.409Z SwingSet: kernel: v4 online: xs-worker static name=uploads transcript entries replayed: 0
2021-07-22T22:17:14.536Z SwingSet: kernel: v6 online: xs-worker static name=vatAdmin transcript entries replayed: 0
2021-07-22T22:17:14.664Z SwingSet: kernel: v8 online: xs-worker static name=vattp transcript entries replayed: 0
2021-07-22T22:17:15.175Z start: adding HTTP/WS listener on 127.0.0.1:7999
2021-07-22T22:17:15.178Z web: Serving static files from /private/var/folders/g1/crzbwthn3n72s_8smhth9g_c0000gn/T/startsolo.83908/html
2021-07-22T22:17:15.180Z start: adding follower/sender for fake chain mySimGCI
2021-07-22T22:17:15.182Z launch-chain: Launching SwingSet kernel
2021-07-22T22:17:15.202Z web: Listening on 127.0.0.1:7999
.
2021-07-22T22:17:15.212Z web: 127.0.0.1:64774[1]: new WebSocket /private/captp?accessToken=8nxqnCoDMMxOHcbmG6yainDt786vw-23LjCOPMv3yrXQs9MxG_Oj54U1PdmhMZ23
o
o
2021-07-22T22:17:24.716Z SwingSet: kernel: v1 online: xs-worker static name=bank transcript entries replayed: 0
2021-07-22T22:17:24.880Z SwingSet: kernel: v2 online: xs-worker static name=board transcript entries replayed: 0
2021-07-22T22:17:25.470Z SwingSet: kernel: v11 online: xs-worker static name=bootstrap transcript entries replayed: 0
2021-07-22T22:17:25.485Z SwingSet: kernel: v13 online: local static name=comms transcript entries replayed: null
2021-07-22T22:17:25.625Z SwingSet: kernel: v3 online: xs-worker static name=distributeFees transcript entries replayed: 0
2021-07-22T22:17:25.796Z SwingSet: kernel: v4 online: xs-worker static name=ibc transcript entries replayed: 0
2021-07-22T22:17:25.974Z SwingSet: kernel: v5 online: xs-worker static name=mints transcript entries replayed: 0
2021-07-22T22:17:26.127Z SwingSet: kernel: v6 online: xs-worker static name=network transcript entries replayed: 0
2021-07-22T22:17:26.345Z SwingSet: kernel: v7 online: xs-worker static name=priceAuthority transcript entries replayed: 0
2021-07-22T22:17:26.481Z SwingSet: kernel: v8 online: xs-worker static name=provisioning transcript entries replayed: 0
2021-07-22T22:17:26.616Z SwingSet: kernel: v9 online: xs-worker static name=sharing transcript entries replayed: 0
2021-07-22T22:17:26.759Z SwingSet: kernel: v15 online: xs-worker static name=timer transcript entries replayed: 0
2021-07-22T22:17:26.884Z SwingSet: kernel: v12 online: xs-worker static name=vatAdmin transcript entries replayed: 0
2021-07-22T22:17:27.020Z SwingSet: kernel: v14 online: xs-worker static name=vattp transcript entries replayed: 0
2021-07-22T22:17:27.349Z SwingSet: kernel: v10 online: xs-worker static name=zoe transcript entries replayed: 0
2021-07-22T22:17:28.272Z SwingSet: kernel: v16 online: xs-worker  transcript entries replayed: 0
2021-07-22T22:17:28.535Z SwingSet: kernel: v17 online: xs-worker  transcript entries replayed: 0
2021-07-22T22:17:29.604Z SwingSet: kernel: v18 online: xs-worker  transcript entries replayed: 0
Logging sent error stack (Error#1)
Error#1: Bank doesn't implement fee collectors
Error: Bank doesn't implement fee collectors
 at apply ()
 at Error ()
 at getFeeCollectorDepositFacet ()
 at ()
 at win ()
 at ()

Error#1 ERROR_NOTE: Thrown from: (Error#2) : 2 . 0
Error#1 ERROR_NOTE: Sent as error:liveSlots:v1#70001
Error#2: Event: 1.1
Error: Event: 1.1
 at construct ()
 at Error ()
 at trackTurns ()
 at handle ()
 at deliver ()
 at dispatchToUserspace ()
 at ()

2021-07-22T22:17:30.352Z SwingSet: vat: v11: Cannot create fee collector [Error: Bank doesn't implement fee collectors]
2021-07-22T22:17:31.708Z SwingSet: kernel: v19 online: xs-worker  transcript entries replayed: 0
2021-07-22T22:17:31.979Z SwingSet: kernel: v20 online: xs-worker  transcript entries replayed: 0
2021-07-22T22:17:32.273Z SwingSet: kernel: v21 online: xs-worker  transcript entries replayed: 0
2021-07-22T22:17:32.542Z SwingSet: kernel: v22 online: xs-worker  transcript entries replayed: 0
2021-07-22T22:17:32.804Z SwingSet: kernel: v23 online: xs-worker  transcript entries replayed: 0
2021-07-22T22:17:39.415Z SwingSet: vat: v16: ----- ST.5  2 depositValue {
  Central: { brand: '[Alleged: RUN brand]', value: '[0n]' },
  Liquidity: {
    brand: '[Alleged: BLDLiquidity brand]',
    value: '[558000000000000n]'
  },
  Secondary: { brand: '[Alleged: BLD brand]', value: '[0n]' }
}
2021-07-22T22:17:39.463Z SwingSet: vat: v16: ----- ST.5  3 depositValue {
  Central: { brand: '[Alleged: RUN brand]', value: '[0n]' },
  Liquidity: {
    brand: '[Alleged: ATOMLiquidity brand]',
    value: '[18610000000000n]'
  },
  Secondary: { brand: '[Alleged: ATOM brand]', value: '[0n]' }
}
2021-07-22T22:17:39.500Z SwingSet: vat: v16: ----- ST.5  4 depositValue {
  Central: { brand: '[Alleged: RUN brand]', value: '[0n]' },
  Liquidity: {
    brand: '[Alleged: ETHLiquidity brand]',
    value: '[1914860000000000n]'
  },
  Secondary: { brand: '[Alleged: ETH brand]', value: '[0n]' }
}
2021-07-22T22:17:39.539Z SwingSet: vat: v16: ----- ST.5  5 depositValue {
  Central: { brand: '[Alleged: RUN brand]', value: '[0n]' },
  Liquidity: {
    brand: '[Alleged: LINKLiquidity brand]',
    value: '[27900000000000n]'
  },
  Secondary: { brand: '[Alleged: LINK brand]', value: '[0n]' }
}
2021-07-22T22:17:39.576Z SwingSet: vat: v16: ----- ST.5  6 depositValue {
  Central: { brand: '[Alleged: RUN brand]', value: '[0n]' },
  Liquidity: {
    brand: '[Alleged: USDCLiquidity brand]',
    value: '[10000000000000n]'
  },
  Secondary: { brand: '[Alleged: USDC brand]', value: '[0n]' }
}
2021-07-22T22:17:40.342Z SwingSet: vat: v11: Making fake price authority for RUN-moola
2021-07-22T22:17:40.343Z SwingSet: vat: v11: Making fake price authority for moola-RUN
2021-07-22T22:17:40.353Z SwingSet: vat: v11: Making fake price authority for RUN-simolean
2021-07-22T22:17:40.354Z SwingSet: vat: v11: Making fake price authority for simolean-RUN
2021-07-22T22:17:40.736Z start: swingset running
'@agoric/install-ses' sniffed and found a 'LOCKDOWN_OPTIONS' environment variable

Open CapTP connection to ws://127.0.0.1:7999/private/captp...
2021-07-22T22:17:42.289Z web: 127.0.0.1:64779[2]: new WebSocket /private/captp?accessToken=8nxqnCoDMMxOHcbmG6yainDt786vw-23LjCOPMv3yrXQs9MxG_Oj54U1PdmhMZ23
o
o
o
o
[... many many more]
o
o
o
o
o
^C  
  ✖ Exiting due to SIGINT

  6 tests were pending in test/test-home.js

  ◌ home.board
  ◌ home.wallet - receive zoe invite
  ◌ home.wallet - central issuer setup
  ◌ home.localTimerService makeNotifier
  ◌ home.localTimerService createRepeater
  ◌ home.localTimerService makeRepeater

  ─

solo(2018-markm-far-must-be-declared)$ 
```

**Comment by michaelfig** [comment]
> (1):solo(2018-markm-far-must-be-declared)$ NOISY=true yarn test -v

All those `o`s are the sim-chain reporting that the CapTP connection is ready, but somehow the messages from the solo are being silently dropped (or at least, not resulting in fulfilment of an `E(x)` promise).  Probably the same "notifier never fires" problem.

File: packages/marshal/src/passStyleOf.js [hunk 1]
```diff
@@ -236,17 +236,23 @@ export { assertIface };
...
     !(
       check(
         typeof val === 'object',
         X`cannot serialize non-objects like ${val}`,
       ) &&
       check(!Array.isArray(val), X`Arrays cannot be pass-by-remote`) &&
-      check(val !== null, X`null cannot be pass-by-remote`)
+      check(val !== null, X`null cannot be pass-by-remote`) &&
+      check(
+        // @ts-ignore
```
Comment (michaelfig) [inline packages/marshal/src/passStyleOf.js]: Please write the reason for the ts-ignore on the same line:
```suggestion
        // @ts-ignore - this is the reason we needed to ignore
```

File: packages/marshal/src/passStyleOf.js [hunk 2]
```diff
@@ -362,9 +368,15 @@ function checkRemotable(val, check = x => x) {
   const p = getPrototypeOf(val);
 
   if (ALLOW_IMPLICIT_REMOTABLES && (p === null || p === objectPrototype)) {
+    const err = assert.error(
+      X`Remotables should be explicitly declared: ${q(val)}`,
+    );
+    console.warn('Missing Far:', err);
+    // eslint-disable-next-line no-debugger
+    debugger;
```
Comment (michaelfig) [inline packages/marshal/src/passStyleOf.js]: I don't know the exact circumstances: will this turn on the deoptimiser?

**Review by michaelfig — APPROVED** [review]
Looks good!

File: packages/marshal/src/passStyleOf.js (hunk 1 above)
Comment (erights) [reply packages/marshal/src/passStyleOf.js]: Done

File: packages/marshal/src/passStyleOf.js (hunk 2 above)
Comment (erights) [reply packages/marshal/src/passStyleOf.js]: Sorry, wasn't meant for production. Gone.

**Pull request merged**